huaweicloud/services/cce/resource_huaweicloud_cce_node.go

```diff
@@ -1102,118 +1110,151 @@ func resourceNodeUpdate(ctx context.Context, d *schema.ResourceData, meta interf
 	if err != nil {
 		return diag.Errorf("error creating CCE client: %s", err)
 	}
 
 	clusterid := d.Get("cluster_id").(string)
 	// remove node without deleting ecs
 	if d.Get("keep_ecs").(bool) {
-		loginSpec, err := buildResourceNodeLoginpec(d)
+		err := removeNode(nodeClient, d, clusterid)
 		if err != nil {
-			diag.FromErr(err)
+			return diag.FromErr(err)
+		}
+	} else {
+		err := deleteNode(ctx, cfg, nodeClient, d, clusterid)
+		if err != nil {
+			return diag.FromErr(err)
 		}
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"Deleting"},
+		Target:       []string{"Deleted"},
+		Refresh:      nodeStateRefreshFunc(nodeClient, clusterid, d.Id()),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        60 * time.Second,
+		PollInterval: 20 * time.Second,
+	}
 
-		removeOpts := nodes.RemoveOpts{
-			Spec: nodes.RemoveNodeSpec{
-				Login: loginSpec,
-				Nodes: []nodes.NodeItem{
-					{
-						Uid: d.Id(),
-					},
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("error deleting CCE Node: %s", err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func removeNode(nodeClient *golangsdk.ServiceClient, d *schema.ResourceData, clusterID string) error {
+	loginSpec, err := buildResourceNodeLoginSpec(d)
+	if err != nil {
+		return err
+	}
+
+	removeOpts := nodes.RemoveOpts{
+		Spec: nodes.RemoveNodeSpec{
+			Login: loginSpec,
+			Nodes: []nodes.NodeItem{
+				{
+					Uid: d.Id(),
 				},
 			},
-		}
+		},
+	}
+
+	err = nodes.Remove(nodeClient, clusterID, removeOpts).ExtractErr()
+	if err != nil {
+		return fmt.Errorf("error removing CCE node: %s", err)
+	}
+
+	return nil
+}
 
-		err = nodes.Remove(nodeClient, clusterid, removeOpts).ExtractErr()
+func deleteNode(ctx context.Context, cfg *config.Config, nodeClient *golangsdk.ServiceClient,
+	d *schema.ResourceData, clusterID string) error {
+	// for prePaid node, firstly, we should unsubscribe the ecs server, and then delete it
+	if d.Get("charging_mode").(string) == "prePaid" || d.Get("billing_mode").(int) == 1 {
+		region := cfg.GetRegion(d)
+		serverID := d.Get("server_id").(string)
+		publicIP := d.Get("public_ip").(string)
+
+		resourceIDs, err := getResourceIDsToUnsubscribe(cfg, d, serverID, publicIP)
 		if err != nil {
-			return diag.Errorf("error removing CCE node: %s", err)
+			return err
+		}
+		computeClient, err := cfg.ComputeV1Client(region)
+		if err != nil {
+			return fmt.Errorf("error creating compute client: %s", err)
 		}
-	} else {
-		// for prePaid node, firstly, we should unsubscribe the ecs server, and then delete it
-		if d.Get("charging_mode").(string) == "prePaid" || d.Get("billing_mode").(int) == 1 {
-			serverID := d.Get("server_id").(string)
-			publicIP := d.Get("public_ip").(string)
 
-			resourceIDs := make([]string, 0, 2)
-			computeClient, err := config.ComputeV1Client(config.GetRegion(d))
-			if err != nil {
-				return diag.Errorf("error creating compute client: %s", err)
+		if len(resourceIDs) > 0 {
```

Comment:
it's better to judge the length of resourceIDs immediately after getting the value.

Reply:
sure